Update the tutorial for Rails (#1258)
claviska/shoelace-css: docs/tutorials/integrating-with-rails.md

- Rails >= 6
- Node >= 12.10
- Yarn >= 1.22

## Instructions

When using Shoelace, there are mostly three things that need to be served to the client browser:

- Javascript files for the Web Components
- CSS files for light and dark themes (they can co-exist, but one of them is required)
- Shoelace Icons

Depending on the JS bundler you are using, you may need to do some additional configuration. However, the basic steps
should be just about the same. Also, it is recommended to read to the [Bundling section in the Installation](/getting-started/installation?id=bundling)

> to understand how Shoalace can be set up with a JS bundler in general. In this tutorial, we will assume that your Rails

Reviewer: ```suggestion
to understand how Shoelace can be set up with a JS bundler in general. In this tutorial, we will assume that your Rails app
```

> is already set up with a JS bundler that supports importing CSS files directly (e.g. Turbopack, esbuild, Vite).

Reviewer: Is it worth documenting importmaps `<link>` to the CDN for stylesheets?

Author: I have just looked at importmaps, and wow, this seems like anther giant beast. This will take a bit more time to write up a doc for it...

Author: I took a deep dive into the `importmaps-rails` gem and now https://toprubycompanies.info/ runs with it. Overall, it's working great. There is a couple of changes needed in the autoloader, and I will send a separate PR for that.

<img width="939" alt="importmap" src="https://user-images.githubusercontent.com/386234/235344611-b5a92cdb-a109-41bb-b383-c41045ac059f.png">

To get started using Shoelace with Rails, the following package must be installed.

```bash
yarn add @shoelace-style/shoelace
```

> This is required regardless of the JS bunlder you are using.

Reviewer: ```suggestion
This is required regardless of the JS bundler you are using.
```

### Javascript & CSS

The next step is to import the JavaScript files and default theme for Shoelace. Add the following code to your
entrypoint JS file (generally `application.js`).

```js
// application.js
import '@shoelace-style/shoelace';

> // You can also add these two if the JS bundler of your choise supports importing CSS files.

Reviewer 2: choise -> choice

Author: ```suggestion
// You can also add these two if the JS bundler of your choice supports importing CSS files.
```

import '@shoelace-style/shoelace/dist/themes/light.css';
import '@shoelace-style/shoelace/dist/themes/dark.css'; // Optional dark mode
```

> !> In this example, the entire dependencies are imported for simplisity. However, importing directly from

Reviewer: ```suggestion
!> In this example, all Shoelace components are imported for simplicity. However, importing directly from
```

`@shoelace-style/shoelace` may result in a larger bundle size than necessary. Consider importing only the components
you actually need in the actual application.

> You can also import the CSS if you prefer to maintain a separate entrypoint CSS file that is complied

Reviewer: ```suggestion
You can also import the CSS inside of a `.css` file if you prefer to maintain a separate CSS entrypoint.
```

> independently.

Reviewer: ```suggestion
Such as with CssBundling-Rails.
```

// application.css
@import '@shoelace-style/shoelace/dist/themes/light.css';
```

For more details about themes, please refer to [Theme Basics](/getting-started/themes?id=theme-basics).

### Serving up Shoelace Icons

#### Using the `shoelace-rails` gem

You do not have to do anything else if you are using [the `shoelace-rails` gem](https://github.com/yuki24/shoelace-rails).
Here are how it works in different environments:

* In development and test, the icons are served by the `ActionDispatch::Static` middleware, directly from the
  `node_modules/@shoelace-style/shoelace/dist/assets/icons` directory.
* In production, the icon files are automatically copied into the `public/assets` directory as part of the
  `assets:precompile` rake task.

#### Copying Icon files with a Rake task

```ruby
# Rakefile
namespace :shoelace do
  namespace :icons do
    desc "Copy Shoelace icons to the assets path"
    task copy: :environment do
      cp_r "node_modules/@shoelace-style/shoelace/dist/assets", Rails.public_path
    end
  end
end

Rake::Task["assets:precompile"].enhance(["shoelace:icons:copy"])
```

Now you can start using Shoelace components with Rails!

## Using Rails View Helpers

> 🚧

Author: This is still WIP. I'll come back to it some time later this week.
